src/swap/SwapScreenV2.tsx

```diff
@@ -0,0 +1,1149 @@
+    crossChainFee,
+    quote,
+    confirmSwapFailed,
+    priceImpactWarningThreshold,
+  ])
+
+  const allowSwap = useMemo(
+    () =>
+      !warnings.showDecreaseSpendForGasWarning &&
+      !warnings.showNotEnoughBalanceForGasWarning &&
+      !warnings.showInsufficientBalanceWarning &&
+      !warnings.showCrossChainFeeWarning &&
+      !confirmSwapIsLoading &&
+      !shouldShowSkeletons &&
+      processedAmountsFrom.token.bignum &&
+      processedAmountsFrom.token.bignum.gt(0) &&
+      processedAmountsTo.token.bignum &&
+      processedAmountsTo.token.bignum.gt(0),
+    [
+      processedAmountsFrom.token.bignum,
+      processedAmountsTo.token.bignum,
+      shouldShowSkeletons,
+      confirmSwapIsLoading,
+      warnings.showInsufficientBalanceWarning,
+      warnings.showDecreaseSpendForGasWarning,
+      warnings.showNotEnoughBalanceForGasWarning,
+      warnings.showCrossChainFeeWarning,
+    ]
+  )
+
+  useEffect(
+    function refreshTransactionQuote() {
+      setStartedSwapId(undefined)
+      if (!processedAmountsFrom.token.bignum) {
+        clearQuote()
+        replaceAmountTo('')
+        return
+      }
+
+      const debounceTimeout = setTimeout(() => {
+        const bothTokensPresent = !!(fromToken && toToken)
+        const amountIsTooSmall =
+          !processedAmountsFrom.token.bignum || processedAmountsFrom.token.bignum.lte(0)
+
+        if (!bothTokensPresent || amountIsTooSmall) {
+          return
+        }
+
+        // This variable prevents the quote from needlessly being fetched again.
+        const quoteIsTheSameAsTheLastOne =
+          quote &&
+          quote.toTokenId === toToken.tokenId &&
+          quote.fromTokenId === fromToken.tokenId &&
+          processedAmountsFrom.token.bignum &&
+          quote.swapAmount.eq(processedAmountsFrom.token.bignum)
+
+        if (!quoteIsTheSameAsTheLastOne) {
+          replaceAmountTo('')
+          void refreshQuote(
+            fromToken,
+            toToken,
+            { FROM: processedAmountsFrom.token.bignum, TO: null },
+            Field.FROM
+          )
+        }
+      }, FETCH_UPDATED_TRANSACTIONS_DEBOUNCE_TIME_MS)
+
+      return () => {
+        clearTimeout(debounceTimeout)
+      }
+    },
+    [
+      processedAmountsFrom.token.bignum?.toString(),
+      fromToken,
+      toToken,
+      quote,
+      refreshQuote,
+      // clearQuote,
+      // replaceAmountTo,
+    ]
+  )
+
+  useEffect(function trackSwapScreenOpen() {
+    AppAnalytics.track(SwapEvents.swap_screen_open)
+  }, [])
+
+  useEffect(
+    function trackImpactWarningDisplayed() {
+      if (warnings.showPriceImpactWarning || warnings.showMissingPriceImpactWarning) {
+        if (!quote) {
+          return
+        }
+        const fromToken = tokensById[quote.fromTokenId]
+        const toToken = tokensById[quote.toTokenId]
+
+        if (!fromToken || !toToken) {
+          // Should never happen
+          Logger.error(TAG, 'fromToken or toToken not found')
```

Comment:
if that is type guard (as "should never happen" comment suggests), do we need to log anything?